File: project1/employee/models/crud.py
import mysql.connector
from mysql.connector import Error
from typing import List, Optional, Dict


class CrudOps:

    def __init__(self,
                 host: str = "localhost",
                 user: str = "root",
                 password: str = "1234",
                 database: str = "expense_manager"):

        self.host = host
        self.user = user
        self.password = password
        self.database = database


    def _get_connection(self):
        """Create MySQL connection"""
        return mysql.connector.connect(
            host=self.host,
            user=self.user,
            password=self.password,
            database=self.database
        )

    # ---------- USER OPERATIONS ----------

    def create_user(self, username: str, password: str, role: str = "Employee") -> bool:
        """Creates a new user account"""
        try:
            conn = self._get_connection()
            cursor = conn.cursor(dictionary=True)

            cursor.execute(
                """
                INSERT INTO users (username, password, role)
                VALUES (%s, %s, %s)
                """,
                (username, password, role)
            )

            conn.commit()
            return True

        except Error:
            return False

        finally:
            if conn.is_connected():
                cursor.close()
                conn.close()


    def get_user_by_credentials(self, username: str, password: str) -> Optional[Dict]:
        """Used for login"""
        conn = self._get_connection()
        cursor = conn.cursor(dictionary=True)

        cursor.execute(
            """
            SELECT id, username, role
            FROM users
            WHERE username = %s AND password = %s
            """,
            (username, password)
        )

        row = cursor.fetchone()

        cursor.close()
        conn.close()
        return row


    def get_user_by_id(self, user_id: int) -> Optional[Dict]:
        conn = self._get_connection()
        cursor = conn.cursor(dictionary=True)

        cursor.execute(
            "SELECT * FROM users WHERE id = %s",
            (user_id,)
        )

        row = cursor.fetchone()

        cursor.close()
        conn.close()
        return row


    def get_all_users(self) -> List[Dict]:
        conn = self._get_connection()
        cursor = conn.cursor(dictionary=True)

        cursor.execute("SELECT * FROM users ORDER BY username")

        results = cursor.fetchall()

        cursor.close()
        conn.close()

        return results

    # ---------- EXPENSE OPERATIONS ----------

    def create_expense(self, user_id: int, amount: float, description: str, date: str) -> int:
        """Create expense + default pending approval"""
        conn = self._get_connection()
        cursor = conn.cursor()

        cursor.execute(
            """
            INSERT INTO expenses (user_id, amount, description, date)
            VALUES (%s, %s, %s, %s)
            """,
            (user_id, amount, description, date)
        )

        expense_id = cursor.lastrowid

        cursor.execute(
            """
            INSERT INTO approvals (expense_id, status)
            VALUES (%s, 'pending')
            """,
            (expense_id,)
        )

        conn.commit()
        cursor.close()
        conn.close()

        return expense_id


    def get_expense_by_id(self, expense_id: int) -> Optional[Dict]:
        conn = self._get_connection()
        cursor = conn.cursor(dictionary=True)

        cursor.execute(
            """
            SELECT e.*, a.status, a.comment
            FROM expenses e
            LEFT JOIN approvals a ON e.id = a.expense_id
            WHERE e.id = %s
            """,
            (expense_id,)
        )

        row = cursor.fetchone()

        cursor.close()
        conn.close()

        return row


    def get_expenses_by_user_id(self, user_id: int) -> List[Dict]:
        conn = self._get_connection()
        cursor = conn.cursor(dictionary=True)

        cursor.execute(
            """
            SELECT e.*, a.status
            FROM expenses e
            LEFT JOIN approvals a ON e.id = a.expense_id
            WHERE e.user_id = %s
            ORDER BY e.date DESC
            """,
            (user_id,)
        )

        results = cursor.fetchall()

        cursor.close()
        conn.close()
        return results


    # def get_all_expenses(self) -> List[Dict]:
    #     conn = self._get_connection()
    #     cursor = conn.cursor(dictionary=True)
    #
    #     cursor.execute(
    #         """
    #         SELECT e.*, u.username, a.status, a.comment, a.reviewer, a.review_date
    #         FROM expenses e
    #         JOIN users u ON e.user_id = u.id
    #         LEFT JOIN approvals a ON e.id = a.expense_id
    #         ORDER BY e.date DESC
    #         """
    #     )

    # results = cursor.fetchall()
    #
    # cursor.close()
    # conn.close()
    # return results



    def get_all_expenses_history_by_user_id(self, user_id: int) -> List[Dict]:
        conn = self._get_connection()
        cursor = conn.cursor(dictionary=True)

        cursor.execute(
            """
            SELECT e.*, a.status, a.comment, a.reviewer, a.review_date
            FROM expenses e
            LEFT JOIN approvals a ON e.id = a.expense_id
            WHERE e.user_id = %s
            ORDER BY e.date DESC
            """,
            (user_id,)
        )

        results = cursor.fetchall()

        cursor.close()
        conn.close()
        return results





    def update_expense(self, expense_id: int, amount: float, description: str) -> bool:
        conn = self._get_connection()
        cursor = conn.cursor()

        cursor.execute(
            """
            UPDATE expenses
            SET amount = %s, description = %s
            WHERE id = %s
            """,
            (amount, description, expense_id)
        )

        conn.commit()
        updated = cursor.rowcount > 0

        cursor.close()
        conn.close()

        return updated


    # ---------- APPROVAL OPERATIONS ----------

    def get_approval_by_expense_id(self, expense_id: int) -> Optional[Dict]:
        conn = self._get_connection()
        cursor = conn.cursor(dictionary=True)

        cursor.execute(
            """
            SELECT *
            FROM approvals
            WHERE expense_id = %s
            """,
            (expense_id,)
        )

        row = cursor.fetchone()

        cursor.close()
        conn.close()

        return row


    def update_approval_status(self,
                               expense_id: int,
                               status: str,
                               reviewer_id: int = None,
                               comment: str = None) -> bool:

        conn = self._get_connection()
        cursor = conn.cursor()

        cursor.execute(
            """
            UPDATE approvals
            SET status = %s,
                reviewer = %s,
                comment = %s,
                review_date = CURDATE()
            WHERE expense_id = %s
            """,
            (status, reviewer_id, comment, expense_id)
        )

        conn.commit()
        updated = cursor.rowcount > 0

        cursor.close()
        conn.close()

        return updated


    def get_pending_approvals(self) -> List[Dict]:
        conn = self._get_connection()
        cursor = conn.cursor(dictionary=True)

        cursor.execute(
            """
            SELECT e.*, u.username, a.id AS approval_id
            FROM expenses e
            JOIN users u ON e.user_id = u.id
            JOIN approvals a ON e.id = a.expense_id
            WHERE a.status = 'pending'
            ORDER BY e.date DESC
            """
        )

        results = cursor.fetchall()

        cursor.close()
        conn.close()

        return results

    #
    def get_pending_expenses_by_user_id(self, user_id: int) -> List[Dict]:
        conn = self._get_connection()
        cursor = conn.cursor(dictionary=True)

        cursor.execute(
            """
            SELECT e.*, a.status
            FROM expenses e
            JOIN approvals a ON e.id = a.expense_id
            WHERE e.user_id = %s AND a.status = 'pending'
            ORDER BY e.date DESC
            """,
            (user_id,)
        )

        results = cursor.fetchall()

        cursor.close()
        conn.close()
        return results



    # ---------- DELETE ----------

    def delete_expense(self, expense_id: int) -> bool:
        try:
            conn = self._get_connection()
            cursor = conn.cursor()

            cursor.execute("DELETE FROM approvals WHERE expense_id = %s", (expense_id,))
            cursor.execute("DELETE FROM expenses WHERE id = %s", (expense_id,))

            conn.commit()
            return cursor.rowcount > 0

        except Error:
            return False

        finally:
            if conn.is_connected():
                cursor.close()
                conn.close()

File: project1/manage/src/main/java/com/raveture/manage/model/CRUD.java
package com.raveture.manage.model;

import java.sql.*;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;

public class CRUD {

    private final String host;
    private final String user;
    private final String password;
    private final String database;

    public CRUD(String host, String user, String password, String database) {
        this.host = host;
        this.user = user;
        this.password = password;
        this.database = database;
    }

   private Connection getConnection() throws SQLException {
    try {
        Class.forName("com.mysql.cj.jdbc.Driver");
    } catch (ClassNotFoundException e) {
        e.printStackTrace();
    }

    String url = "jdbc:mysql://" + host + ":3306/" + database + "?useSSL=false&serverTimezone=UTC";
    return DriverManager.getConnection(url, user, password);
}


    // ---------- USER OPERATIONS ----------
    public Map<String, Object> getUserById(int userId) {
        String sql = "SELECT id, username, role FROM users WHERE id = ?";
        try (Connection conn = getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setInt(1, userId);
            ResultSet rs = stmt.executeQuery();
            if (rs.next()) {
                Map<String, Object> user = new HashMap<>();
                user.put("id", rs.getInt("id"));
                user.put("username", rs.getString("username"));
                user.put("role", rs.getString("role"));
                return user;
            }

        } catch (SQLException e) {
            e.printStackTrace();
        }
        return null;
    }

    public Map<String, Object> getUserByCredentials(String username, String password) {
        String sql = "SELECT id, username, role FROM users WHERE username = ? AND password = ?";
        try (Connection conn = getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setString(1, username);
            stmt.setString(2, password);
            ResultSet rs = stmt.executeQuery();
            if (rs.next()) {
                Map<String, Object> user = new HashMap<>();
                user.put("id", rs.getInt("id"));
                user.put("username", rs.getString("username"));
                user.put("role", rs.getString("role"));
                return user;
            }

        } catch (SQLException e) {
            e.printStackTrace();
        }
        return null;
    }

    // ---------- EXPENSE OPERATIONS ----------
    public List<Map<String, Object>> getAllExpenses() {
        String sql = "SELECT e.id, e.user_id, e.amount, e.description, e.date, a.status, a.comment " +
                     "FROM expenses e JOIN approvals a ON e.id = a.expense_id ORDER BY e.date DESC";
        return executeQuery(sql);
    }

    public List<Map<String, Object>> getExpensesByStatus(String status) {
        String sql = "SELECT e.id, e.user_id, e.amount, e.description, e.date, a.status, a.comment " +
                     "FROM expenses e JOIN approvals a ON e.id = a.expense_id WHERE a.status = ? ORDER BY e.date DESC";
        return executeQuery(sql, status);
    }

    public List<Map<String, Object>> getExpensesByEmpId(int empId) {
        String sql = "SELECT e.id, e.user_id, e.amount, e.description, e.date, a.status, a.comment " +
                     "FROM expenses e JOIN approvals a ON e.id = a.expense_id WHERE e.user_id = ? ORDER BY e.date DESC";
        return executeQuery(sql, empId);
    }

    public List<Map<String, Object>> getExpensesByDateRange(String start, String end) {
        String sql = "SELECT e.id, e.user_id, e.amount, e.description, e.date, a.status, a.comment " +
                     "FROM expenses e JOIN approvals a ON e.id = a.expense_id WHERE e.date BETWEEN ? AND ? ORDER BY e.date DESC";
        return executeQuery(sql, start, end);
    }

    // ---------- APPROVAL OPERATIONS ----------
    public boolean changeStatus(int expenseId, String status, int reviewerId) {
        String sql = "UPDATE approvals SET status = ?, reviewer = ?, review_date = NOW() WHERE expense_id = ?";
        try (Connection conn = getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setString(1, status);
            stmt.setInt(2, reviewerId);
            stmt.setInt(3, expenseId);
            return stmt.executeUpdate() > 0;

        } catch (SQLException e) {
            e.printStackTrace();
            return false;
        }
    }

    public boolean updateComment(int expenseId, String comment) {
        String sql = "UPDATE approvals SET comment = ? WHERE expense_id = ?";
        try (Connection conn = getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setString(1, comment);
            stmt.setInt(2, expenseId);
            return stmt.executeUpdate() > 0;

        } catch (SQLException e) {
            e.printStackTrace();
            return false;
        }
    }

    // ---------- DELETE ----------
    public boolean deleteExpense(int expenseId) {
        String sqlApproval = "DELETE FROM approvals WHERE expense_id = ?";
        String sqlExpense = "DELETE FROM expenses WHERE id = ?";

        try (Connection conn = getConnection();
             PreparedStatement stmtApproval = conn.prepareStatement(sqlApproval);
             PreparedStatement stmtExpense = conn.prepareStatement(sqlExpense)) {

            stmtApproval.setInt(1, expenseId);
            stmtApproval.executeUpdate();

            stmtExpense.setInt(1, expenseId);
            return stmtExpense.executeUpdate() > 0;

        } catch (SQLException e) {
            e.printStackTrace();
            return false;
        }
    }

    // ---------- PRIVATE HELPER ----------
    private List<Map<String, Object>> executeQuery(String sql, Object... params) {
        List<Map<String, Object>> list = new ArrayList<>();

        try (Connection conn = getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            for (int i = 0; i < params.length; i++) {
                stmt.setObject(i + 1, params[i]);
            }

            ResultSet rs = stmt.executeQuery();
            ResultSetMetaData meta = rs.getMetaData();
            int colCount = meta.getColumnCount();

            while (rs.next()) {
                Map<String, Object> row = new HashMap<>();
                for (int i = 1; i <= colCount; i++) {
                    row.put(meta.getColumnName(i), rs.getObject(i));
                }
                list.add(row);
            }

        } catch (SQLException e) {
            e.printStackTrace();
        }

        return list;
    }
}
